general/mem_manager.hpp

```diff
@@ -13,33 +13,462 @@
+    operation is based on the enumeration ordering:
+
+        HOST < HOST_32 < HOST_64 < CUDA < CUDA_UVM. */
+MemoryClass operator*(MemoryClass mc1, MemoryClass mc2);
+
+/// Class used by MFEM to store pointers to host and/or device memory.
+/** The template class paramter, T, must be a plain-old-data (POD) type.
+
+    In many respects this class behaves like a pointer:
+    * When destroyed, a Memory object does NOT automatically delete any
+      allocated memory.
+    * Only the method Delete() will deallocate a Memory object.
+    * Other methods that modify the object (e.g. New(), Wrap(), etc) will simply
+      overwrite the old contents.
+    * One difference with a pointer is that a const Memory object is NOT like a
+      const pointer: it cannot be used to overwrite entries in the memory.
+
+    A Memory object stores up to two different pointers: one host pointer (with
+    MemoryType from MemoryClass::HOST) and one device pointer (currently one of
+    MemoryType::CUDA or MemoryTyep::CUDA_UVM).
+
+    A Memory object can hold (wrap) an externally allocated pointer with any
+    given MemoryType.
+
+    Access to the content of the Memory object can be requested with any given
+    MemoryClass through the methods ReadWriteAccess(), ReadAccess(), and
+    WriteAccess(). Requesting such access may result in additional (internally
+    handled) memory allocation and/or memory copy.
+    * When ReadWriteAccess() is called, the returned pointer becomes the only
+      valid pointer.
+    * When ReadAccess() is called, the returned pointer becomes valid, however
+      the other pointer (host or device) may remain valid as well.
+    * When WriteAccess() is called, the returned pointer becomes the only valid
+      pointer, however, unlike ReadWriteAccess(), no memory copy will be
+      performed.
+
+    The host memory (pointer from MemoryClass::HOST) can be accessed through the
+    inline methods: `operator[]()`, `operator*()`, the implicit conversion
+    functions `operator T*()`, `operator const T*()`, and the explicit
+    conversion template functions `operator U*()`,  `operator const U*()` (with
+    any suitable type U). In certain cases, using these methods may have
+    undefined behavior, e.g. if the host pointer is not currently valid. */
+template <typename T>
+class Memory
+{
+protected:
+   friend class MemoryManager;
+   friend void MemoryPrintFlags(unsigned flags);
+
+   enum FlagMask
+   {
+      REGISTERED    = 1,   ///< #h_ptr is registered with the MemoryManager
+      OWNS_HOST     = 2,   ///< The host pointer will be deleted by Delete()
+      OWNS_DEVICE   = 4,   ///< The device pointer will be deleted by Delete()
+      OWNS_INTERNAL = 8,   ///< Ownership flag for internal Memory data
+      VALID_HOST    = 16,  ///< Host pointer is valid
+      VALID_DEVICE  = 32,  ///< Device pointer is valid
+      ALIAS         = 64,
+      EXEC_FLAG     = 128  ///< Optional execution flag used by MFEM
```

Comment:
What's an execution flag?

Reply:
I get it now, maybe a reference like "See XX" would be good.

Reply:
I think the word "optional" is not precise -- I'll replace the text with "Execution flag reserved for use by MFEM". What would you suggest for reference? Maybe "e.g. see `Vector::UseDevice`"?